llama-cpp-2/src/llama_batch.rs

```diff
@@ -90,7 +90,50 @@ impl LlamaBatch {
 
         Ok(())
     }
-    /// Create a new `LlamaBatch` that cab contain up to `n_tokens` tokens.
+
+    /// Add a sequence of tokens to the batch for the given sequence id. If [logits_all] is true, the
+    /// tokens will be initialized and can be read from after the next decode.
+    ///
+    /// Either way the last token in the sequence will have its logits set to `true`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if there is insufficient space in the buffer
+    pub fn add_sequence(&mut self, tokens: &[LlamaToken],
+                        seq_id: i32,
+                        logits_all: bool) -> Result<(), BatchAddError> {
+        let n_tokens_0 = self.llama_batch.n_tokens;
+        let n_tokens = tokens.len();
+
+        if self.allocated < n_tokens_0 as usize + n_tokens {
+            return Err(BatchAddError::InsufficientSpace(self.allocated));
+        }
+        if n_tokens == 0 {
+            return Ok(())
+        }
+
+        self.llama_batch.n_tokens += n_tokens as i32;
+        for (i, token) in tokens.iter().enumerate() {
+            let j = n_tokens_0 as usize + i;
+            unsafe {
+                self.llama_batch.token.add(j).write(token.0);
+                self.llama_batch.pos.add(j).write(i as i32);
+                let seq_id_ptr = *self.llama_batch.seq_id.add(j);
+                seq_id_ptr.write(seq_id);
+                self.llama_batch.n_seq_id.add(j).write(1);
+
+                let write_logits = logits_all || i == n_tokens - 1;
+                self.llama_batch.logits.add(j).write(write_logits as i8);
+                if write_logits {
+                    self.initialized_logits.push(j as i32);
+                }
+            }
```

Comment:
Yeah, can do. Doing it this way was to avoid redundant checks as much as possible (and avoid the potentially expensive `Vec::retain` call in `LLamaBatch::add`), but it might be a preemptive optimisation.

Alternatively we could extract the unsafe parts into a private function and just share that call between `add` and `add_sequence`, if you'd prefer that?